ompi/runtime/ompi_spc.c

```diff
@@ -339,86 +842,232 @@ void ompi_spc_init(void)
+                    bin_offset += num_bins*2;
+                    continue;
+                }
+                /* This is a non-zero normal counter */
+                opal_output(0, "%s -> %lld\n", ompi_spc_events_names[i].counter_name, recv_buffer[offset+i]);
             }
             opal_output(0, "\n");
-            offset += OMPI_SPC_NUM_COUNTERS;
+            offset += buffer_len;
+            bin_offset += OMPI_SPC_NUM_COUNTERS;
         }
-        printf("###########################################################################\n");
-        printf("NOTE: Any counters not shown here were either disabled or had a value of 0.\n");
-        printf("###########################################################################\n");
+        opal_output(0, "###########################################################################\n");
+        opal_output(0, "NOTE: Any counters not shown here were either disabled or had a value of 0.\n");
+        opal_output(0, "###########################################################################\n");
 
-        free(recv_buffer);
+        free(recv_buffer); recv_buffer = NULL;
     }
-    free(send_buffer);
+    free(send_buffer); send_buffer = NULL;
 
     ompi_spc_comm->c_coll->coll_barrier(ompi_spc_comm, ompi_spc_comm->c_coll->coll_barrier_module);
 }
 
 /* Frees any dynamically alocated OMPI SPC data structures */
 void ompi_spc_fini(void)
 {
+    int fd, rc;
+    char sm_file[128];
+    char *shm_dir = "/dev/shm";
+    orte_proc_t *pptr;
+
+    int rank = ompi_comm_rank(ompi_spc_comm);
+
     if (SPC_ENABLE == 1 && ompi_mpi_spc_dump_enabled) {
         ompi_spc_dump();
     }
 
-    free(ompi_spc_events); ompi_spc_events = NULL;
-    ompi_comm_free(&ompi_spc_comm);
+    int i;
+    for(i = 0; i < OMPI_SPC_NUM_COUNTERS; i++) {
+        CLEAR_SPC_BIT(ompi_spc_attached_event, i);
+        CLEAR_SPC_BIT(ompi_spc_timer_event, i);
+        CLEAR_SPC_BIT(ompi_spc_bin_event, i);
+        CLEAR_SPC_BIT(ompi_spc_collective_bin_event, i);
+    }
+    if(need_free) {
+        free(ompi_spc_events); ompi_spc_events = NULL;
+    }
+    ompi_comm_free(&ompi_spc_comm); ompi_spc_comm = NULL;
 }
 
 /* Records an update to a counter using an atomic add operation. */
 void ompi_spc_record(unsigned int event_id, ompi_spc_value_t value)
 {
     /* Denoted unlikely because counters will often be turned off. */
     if( OPAL_UNLIKELY(IS_SPC_BIT_SET(ompi_spc_attached_event, event_id)) ) {
-        OPAL_THREAD_ADD_FETCH_SIZE_T(&(ompi_spc_events[event_id].value), value);
+        OPAL_THREAD_ADD_FETCH_SIZE_T(&(ompi_spc_values[event_id]), value);
+    }
+}
+
+/* Records an update to a bin counter using an atomic add operation. */
+void ompi_spc_bin_record(unsigned int event_id, ompi_spc_value_t value)
+{
+    int *rules;
+    ompi_spc_value_t *bins;
+
+    /* Denoted unlikely because counters will often be turned off. */
+    if( OPAL_UNLIKELY(IS_SPC_BIT_SET(ompi_spc_attached_event, event_id)) ) {
```

Comment:
Can we get this check outside the function and into the upper-level macro ? If we can, we can save the function call.

Reply:
Ok, I've added the check into the macros.